# CS771A Assignment 1
\- Yash Gupta (190997)

## Q1. The automobile dataset

In [1]:
# importing libraries
import numpy as np
from numpy import linalg as lin
import pandas as pd
from matplotlib import pyplot as plt

In [2]:
# names of the columns of the dataset
cols = ['symboling', 'normalized-losses', 'make', 'fuel-type', 'aspiration', 'num-of-doors', 'body-style', 'drive-wheels', 'engine-location', 'wheel-base', 'length', 'width', 'height', 'curb-weight', 'engine-type', 'num-of-cylinders', 'engine-size', 'fuel-system', 'bore', 'stroke', 'compression-ratio', 'horsepower', 'peak-rpm', 'city-mpg', 'highway-mpg', 'price']
cols

['symboling',
 'normalized-losses',
 'make',
 'fuel-type',
 'aspiration',
 'num-of-doors',
 'body-style',
 'drive-wheels',
 'engine-location',
 'wheel-base',
 'length',
 'width',
 'height',
 'curb-weight',
 'engine-type',
 'num-of-cylinders',
 'engine-size',
 'fuel-system',
 'bore',
 'stroke',
 'compression-ratio',
 'horsepower',
 'peak-rpm',
 'city-mpg',
 'highway-mpg',
 'price']

In [3]:
# reading the data
data = pd.read_csv('data_q1/imports-85.data', names=cols)
data

symboling normalized-losses         make fuel-type aspiration  \
0            3                 ?  alfa-romero       gas        std   
1            3                 ?  alfa-romero       gas        std   
2            1                 ?  alfa-romero       gas        std   
3            2               164         audi       gas        std   
4            2               164         audi       gas        std   
..         ...               ...          ...       ...        ...   
200         -1                95        volvo       gas        std   
201         -1                95        volvo       gas      turbo   
202         -1                95        volvo       gas        std   
203         -1                95        volvo    diesel      turbo   
204         -1                95        volvo       gas      turbo   

    num-of-doors   body-style drive-wheels engine-location  wheel-base  ...  \
0            two  convertible          rwd           front        88.6  ...   
1            two  convertible          rwd           front        88.6  ...   
2            two    hatchback          rwd           front        94.5  ...   
3           four        sedan          fwd           front        99.8  ...   
4           four        sedan          4wd           front        99.4  ...   
..           ...          ...          ...             ...         ...  ...   
200         four        sedan          rwd           front       109.1  ...   
201         four        sedan          rwd           front       109.1  ...   
202         four        sedan          rwd           front       109.1  ...   
203         four        sedan          rwd           front       109.1  ...   
204         four        sedan          rwd           front       109.1  ...   

     engine-size  fuel-system  bore  stroke compression-ratio horsepower  \
0            130         mpfi  3.47    2.68               9.0        111   
1            130         mpfi  3.47    2.68               9.0        111   
2            152         mpfi  2.68    3.47               9.0        154   
3            109         mpfi  3.19    3.40              10.0        102   
4            136         mpfi  3.19    3.40               8.0        115   
..           ...          ...   ...     ...               ...        ...   
200          141         mpfi  3.78    3.15               9.5        114   
201          141         mpfi  3.78    3.15               8.7        160   
202          173         mpfi  3.58    2.87               8.8        134   
203          145          idi  3.01    3.40              23.0        106   
204          141         mpfi  3.78    3.15               9.5        114   

     peak-rpm city-mpg highway-mpg  price  
0        5000       21          27  13495  
1        5000       21          27  16500  
2        5000       19          26  16500  
3        5500       24          30  13950  
4        5500       18          22  17450  
..        ...      ...         ...    ...  
200      5400       23          28  16845  
201      5300       19          25  19045  
202      5500       18          23  21485  
203      4800       26          27  22470  
204      5400       19          25  22625  

[205 rows x 26 columns]

## Data Preprocessing

### Analysing missing data

In [4]:
# putting NaN for missing values
data.replace('?', np.nan, inplace=True)
data

symboling normalized-losses         make fuel-type aspiration  \
0            3               NaN  alfa-romero       gas        std   
1            3               NaN  alfa-romero       gas        std   
2            1               NaN  alfa-romero       gas        std   
3            2               164         audi       gas        std   
4            2               164         audi       gas        std   
..         ...               ...          ...       ...        ...   
200         -1                95        volvo       gas        std   
201         -1                95        volvo       gas      turbo   
202         -1                95        volvo       gas        std   
203         -1                95        volvo    diesel      turbo   
204         -1                95        volvo       gas      turbo   

    num-of-doors   body-style drive-wheels engine-location  wheel-base  ...  \
0            two  convertible          rwd           front        88.6  ...   
1            two  convertible          rwd           front        88.6  ...   
2            two    hatchback          rwd           front        94.5  ...   
3           four        sedan          fwd           front        99.8  ...   
4           four        sedan          4wd           front        99.4  ...   
..           ...          ...          ...             ...         ...  ...   
200         four        sedan          rwd           front       109.1  ...   
201         four        sedan          rwd           front       109.1  ...   
202         four        sedan          rwd           front       109.1  ...   
203         four        sedan          rwd           front       109.1  ...   
204         four        sedan          rwd           front       109.1  ...   

     engine-size  fuel-system  bore  stroke compression-ratio horsepower  \
0            130         mpfi  3.47    2.68               9.0        111   
1            130         mpfi  3.47    2.68               9.0        111   
2            152         mpfi  2.68    3.47               9.0        154   
3            109         mpfi  3.19    3.40              10.0        102   
4            136         mpfi  3.19    3.40               8.0        115   
..           ...          ...   ...     ...               ...        ...   
200          141         mpfi  3.78    3.15               9.5        114   
201          141         mpfi  3.78    3.15               8.7        160   
202          173         mpfi  3.58    2.87               8.8        134   
203          145          idi  3.01    3.40              23.0        106   
204          141         mpfi  3.78    3.15               9.5        114   

     peak-rpm city-mpg highway-mpg  price  
0        5000       21          27  13495  
1        5000       21          27  16500  
2        5000       19          26  16500  
3        5500       24          30  13950  
4        5500       18          22  17450  
..        ...      ...         ...    ...  
200      5400       23          28  16845  
201      5300       19          25  19045  
202      5500       18          23  21485  
203      4800       26          27  22470  
204      5400       19          25  22625  

[205 rows x 26 columns]

In [5]:
# checking how many missing values are there in each column
data.isnull().sum()

symboling             0
normalized-losses    41
make                  0
fuel-type             0
aspiration            0
num-of-doors          2
body-style            0
drive-wheels          0
engine-location       0
wheel-base            0
length                0
width                 0
height                0
curb-weight           0
engine-type           0
num-of-cylinders      0
engine-size           0
fuel-system           0
bore                  4
stroke                4
compression-ratio     0
horsepower            2
peak-rpm              2
city-mpg              0
highway-mpg           0
price                 4
dtype: int64

In [6]:
# dropping the normalized-losses column as it has a lot of missing data and also does not seem very important
data.drop('normalized-losses', axis=1, inplace=True)
data

symboling         make fuel-type aspiration num-of-doors   body-style  \
0            3  alfa-romero       gas        std          two  convertible   
1            3  alfa-romero       gas        std          two  convertible   
2            1  alfa-romero       gas        std          two    hatchback   
3            2         audi       gas        std         four        sedan   
4            2         audi       gas        std         four        sedan   
..         ...          ...       ...        ...          ...          ...   
200         -1        volvo       gas        std         four        sedan   
201         -1        volvo       gas      turbo         four        sedan   
202         -1        volvo       gas        std         four        sedan   
203         -1        volvo    diesel      turbo         four        sedan   
204         -1        volvo       gas      turbo         four        sedan   

    drive-wheels engine-location  wheel-base  length  ...  engine-size  \
0            rwd           front        88.6   168.8  ...          130   
1            rwd           front        88.6   168.8  ...          130   
2            rwd           front        94.5   171.2  ...          152   
3            fwd           front        99.8   176.6  ...          109   
4            4wd           front        99.4   176.6  ...          136   
..           ...             ...         ...     ...  ...          ...   
200          rwd           front       109.1   188.8  ...          141   
201          rwd           front       109.1   188.8  ...          141   
202          rwd           front       109.1   188.8  ...          173   
203          rwd           front       109.1   188.8  ...          145   
204          rwd           front       109.1   188.8  ...          141   

     fuel-system  bore stroke compression-ratio  horsepower peak-rpm city-mpg  \
0           mpfi  3.47   2.68               9.0         111     5000       21   
1           mpfi  3.47   2.68               9.0         111     5000       21   
2           mpfi  2.68   3.47               9.0         154     5000       19   
3           mpfi  3.19   3.40              10.0         102     5500       24   
4           mpfi  3.19   3.40               8.0         115     5500       18   
..           ...   ...    ...               ...         ...      ...      ...   
200         mpfi  3.78   3.15               9.5         114     5400       23   
201         mpfi  3.78   3.15               8.7         160     5300       19   
202         mpfi  3.58   2.87               8.8         134     5500       18   
203          idi  3.01   3.40              23.0         106     4800       26   
204         mpfi  3.78   3.15               9.5         114     5400       19   

    highway-mpg  price  
0            27  13495  
1            27  16500  
2            26  16500  
3            30  13950  
4            22  17450  
..          ...    ...  
200          28  16845  
201          25  19045  
202          23  21485  
203          27  22470  
204          25  22625  

[205 rows x 25 columns]

In [7]:
# cols_nominal and cols_numerical represent the columns having nominal and numerical data respectively
cols_nominal = ['make', 'fuel-type', 'aspiration', 'num-of-doors', 'body-style', 'drive-wheels', 'engine-location', 'engine-type', 'num-of-cylinders', 'fuel-system']
cols_numerical = ['symboling', 'wheel-base', 'length', 'width', 'height', 'curb-weight', 'engine-size', 'bore', 'stroke', 'compression-ratio', 'horsepower', 'peak-rpm', 'city-mpg', 'highway-mpg', 'price']

In [8]:
# checking the data types of the columns
data.dtypes

symboling              int64
make                  object
fuel-type             object
aspiration            object
num-of-doors          object
body-style            object
drive-wheels          object
engine-location       object
wheel-base           float64
length               float64
width                float64
height               float64
curb-weight            int64
engine-type           object
num-of-cylinders      object
engine-size            int64
fuel-system           object
bore                  object
stroke                object
compression-ratio    float64
horsepower            object
peak-rpm              object
city-mpg               int64
highway-mpg            int64
price                 object
dtype: object

In [9]:
# changing the data types of the numerical columns to numerical data types
for col in cols_numerical:
    data[col] = pd.to_numeric(data[col])
data.dtypes

symboling              int64
make                  object
fuel-type             object
aspiration            object
num-of-doors          object
body-style            object
drive-wheels          object
engine-location       object
wheel-base           float64
length               float64
width                float64
height               float64
curb-weight            int64
engine-type           object
num-of-cylinders      object
engine-size            int64
fuel-system           object
bore                 float64
stroke               float64
compression-ratio    float64
horsepower           float64
peak-rpm             float64
city-mpg               int64
highway-mpg            int64
price                float64
dtype: object

In [10]:
# dropping the rows having missing data for num-of-doors and price
# because num-of-doors is a nominal variable so we can't impute it by finding its average
# and price is the target variable and we don't want to impute the target variable
data.dropna(subset=['num-of-doors', 'price'], inplace=True)
data

symboling         make fuel-type aspiration num-of-doors   body-style  \
0            3  alfa-romero       gas        std          two  convertible   
1            3  alfa-romero       gas        std          two  convertible   
2            1  alfa-romero       gas        std          two    hatchback   
3            2         audi       gas        std         four        sedan   
4            2         audi       gas        std         four        sedan   
..         ...          ...       ...        ...          ...          ...   
200         -1        volvo       gas        std         four        sedan   
201         -1        volvo       gas      turbo         four        sedan   
202         -1        volvo       gas        std         four        sedan   
203         -1        volvo    diesel      turbo         four        sedan   
204         -1        volvo       gas      turbo         four        sedan   

    drive-wheels engine-location  wheel-base  length  ...  engine-size  \
0            rwd           front        88.6   168.8  ...          130   
1            rwd           front        88.6   168.8  ...          130   
2            rwd           front        94.5   171.2  ...          152   
3            fwd           front        99.8   176.6  ...          109   
4            4wd           front        99.4   176.6  ...          136   
..           ...             ...         ...     ...  ...          ...   
200          rwd           front       109.1   188.8  ...          141   
201          rwd           front       109.1   188.8  ...          141   
202          rwd           front       109.1   188.8  ...          173   
203          rwd           front       109.1   188.8  ...          145   
204          rwd           front       109.1   188.8  ...          141   

     fuel-system  bore stroke compression-ratio  horsepower peak-rpm  \
0           mpfi  3.47   2.68               9.0       111.0   5000.0   
1           mpfi  3.47   2.68               9.0       111.0   5000.0   
2           mpfi  2.68   3.47               9.0       154.0   5000.0   
3           mpfi  3.19   3.40              10.0       102.0   5500.0   
4           mpfi  3.19   3.40               8.0       115.0   5500.0   
..           ...   ...    ...               ...         ...      ...   
200         mpfi  3.78   3.15               9.5       114.0   5400.0   
201         mpfi  3.78   3.15               8.7       160.0   5300.0   
202         mpfi  3.58   2.87               8.8       134.0   5500.0   
203          idi  3.01   3.40              23.0       106.0   4800.0   
204         mpfi  3.78   3.15               9.5       114.0   5400.0   

     city-mpg  highway-mpg    price  
0          21           27  13495.0  
1          21           27  16500.0  
2          19           26  16500.0  
3          24           30  13950.0  
4          18           22  17450.0  
..        ...          ...      ...  
200        23           28  16845.0  
201        19           25  19045.0  
202        18           23  21485.0  
203        26           27  22470.0  
204        19           25  22625.0  

[199 rows x 25 columns]

In [11]:
# filling the remaining missing values with the average of the respective columns
data.fillna(data.mean(), inplace=True)
data

symboling         make fuel-type aspiration num-of-doors   body-style  \
0            3  alfa-romero       gas        std          two  convertible   
1            3  alfa-romero       gas        std          two  convertible   
2            1  alfa-romero       gas        std          two    hatchback   
3            2         audi       gas        std         four        sedan   
4            2         audi       gas        std         four        sedan   
..         ...          ...       ...        ...          ...          ...   
200         -1        volvo       gas        std         four        sedan   
201         -1        volvo       gas      turbo         four        sedan   
202         -1        volvo       gas        std         four        sedan   
203         -1        volvo    diesel      turbo         four        sedan   
204         -1        volvo       gas      turbo         four        sedan   

    drive-wheels engine-location  wheel-base  length  ...  engine-size  \
0            rwd           front        88.6   168.8  ...          130   
1            rwd           front        88.6   168.8  ...          130   
2            rwd           front        94.5   171.2  ...          152   
3            fwd           front        99.8   176.6  ...          109   
4            4wd           front        99.4   176.6  ...          136   
..           ...             ...         ...     ...  ...          ...   
200          rwd           front       109.1   188.8  ...          141   
201          rwd           front       109.1   188.8  ...          141   
202          rwd           front       109.1   188.8  ...          173   
203          rwd           front       109.1   188.8  ...          145   
204          rwd           front       109.1   188.8  ...          141   

     fuel-system  bore stroke compression-ratio  horsepower peak-rpm  \
0           mpfi  3.47   2.68               9.0       111.0   5000.0   
1           mpfi  3.47   2.68               9.0       111.0   5000.0   
2           mpfi  2.68   3.47               9.0       154.0   5000.0   
3           mpfi  3.19   3.40              10.0       102.0   5500.0   
4           mpfi  3.19   3.40               8.0       115.0   5500.0   
..           ...   ...    ...               ...         ...      ...   
200         mpfi  3.78   3.15               9.5       114.0   5400.0   
201         mpfi  3.78   3.15               8.7       160.0   5300.0   
202         mpfi  3.58   2.87               8.8       134.0   5500.0   
203          idi  3.01   3.40              23.0       106.0   4800.0   
204         mpfi  3.78   3.15               9.5       114.0   5400.0   

     city-mpg  highway-mpg    price  
0          21           27  13495.0  
1          21           27  16500.0  
2          19           26  16500.0  
3          24           30  13950.0  
4          18           22  17450.0  
..        ...          ...      ...  
200        23           28  16845.0  
201        19           25  19045.0  
202        18           23  21485.0  
203        26           27  22470.0  
204        19           25  22625.0  

[199 rows x 25 columns]

### Feature scaling

In [12]:
# standardizing numerical features except the target variable
for col in cols_numerical[:-1]:
    data[col] = (data[col] - data[col].mean()) / data[col].std()
data

symboling         make fuel-type aspiration num-of-doors   body-style  \
0     1.711409  alfa-romero       gas        std          two  convertible   
1     1.711409  alfa-romero       gas        std          two  convertible   
2     0.123668  alfa-romero       gas        std          two    hatchback   
3     0.917539         audi       gas        std         four        sedan   
4     0.917539         audi       gas        std         four        sedan   
..         ...          ...       ...        ...          ...          ...   
200  -1.464073        volvo       gas        std         four        sedan   
201  -1.464073        volvo       gas      turbo         four        sedan   
202  -1.464073        volvo       gas        std         four        sedan   
203  -1.464073        volvo    diesel      turbo         four        sedan   
204  -1.464073        volvo       gas      turbo         four        sedan   

    drive-wheels engine-location  wheel-base    length  ...  engine-size  \
0            rwd           front   -1.679665 -0.443709  ...     0.070851   
1            rwd           front   -1.679665 -0.443709  ...     0.070851   
2            rwd           front   -0.710243 -0.248952  ...     0.598380   
3            fwd           front    0.160593  0.189252  ...    -0.432699   
4            4wd           front    0.094870  0.189252  ...     0.214723   
..           ...             ...         ...       ...  ...          ...   
200          rwd           front    1.688665  1.179268  ...     0.334616   
201          rwd           front    1.688665  1.179268  ...     0.334616   
202          rwd           front    1.688665  1.179268  ...     1.101930   
203          rwd           front    1.688665  1.179268  ...     0.430530   
204          rwd           front    1.688665  1.179268  ...     0.334616   

     fuel-system      bore    stroke compression-ratio  horsepower  peak-rpm  \
0           mpfi  0.514086 -1.813547         -0.284172    0.197496 -0.246595   
1           mpfi  0.514086 -1.813547         -0.284172    0.197496 -0.246595   
2           mpfi -2.427775  0.675778         -0.284172    1.345735 -0.246595   
3           mpfi -0.528599  0.455205         -0.029120   -0.042834  0.798118   
4           mpfi -0.528599  0.455205         -0.539224    0.304309  0.798118   
..           ...       ...       ...               ...         ...       ...   
200         mpfi  1.668486 -0.332556         -0.156646    0.277605  0.589176   
201         mpfi  1.668486 -0.332556         -0.360688    1.505955  0.380233   
202         mpfi  0.923712 -1.214849         -0.335183    0.811670  0.798118   
203          idi -1.198896  0.455205          3.286562    0.063979 -0.664480   
204         mpfi  1.668486 -0.332556         -0.156646    0.277605  0.589176   

     city-mpg  highway-mpg    price  
0   -0.644520    -0.534152  13495.0  
1   -0.644520    -0.534152  16500.0  
2   -0.956587    -0.681173  16500.0  
3   -0.176420    -0.093089  13950.0  
4   -1.112621    -1.269257  17450.0  
..        ...          ...      ...  
200 -0.332453    -0.387131  16845.0  
201 -0.956587    -0.828194  19045.0  
202 -1.112621    -1.122236  21485.0  
203  0.135647    -0.534152  22470.0  
204 -0.956587    -0.828194  22625.0  

[199 rows x 25 columns]

### Train-test split

In [13]:
# shuffling the rows of the dataframe
data_shuffle = data.sample(frac=1, random_state=1)
data_shuffle

symboling        make fuel-type aspiration num-of-doors body-style  \
181  -1.464073      toyota       gas        std         four      wagon   
42    0.123668       honda       gas        std          two      sedan   
36   -0.670202       honda       gas        std         four      wagon   
95    0.123668      nissan       gas        std          two  hatchback   
151   0.123668      toyota       gas        std          two  hatchback   
..         ...         ...       ...        ...          ...        ...   
139   0.917539      subaru       gas        std          two  hatchback   
143  -0.670202      subaru       gas        std         four      sedan   
77    0.917539  mitsubishi       gas        std          two  hatchback   
146  -0.670202      subaru       gas        std         four      wagon   
39   -0.670202       honda       gas        std         four      sedan   

    drive-wheels engine-location  wheel-base    length  ...  engine-size  \
181          rwd           front    0.932844  1.098119  ...     0.814187   
42           fwd           front   -0.381626 -0.419364  ...    -0.408720   
36           fwd           front   -0.381626 -1.393150  ...    -0.840335   
95           fwd           front   -0.710243 -0.703385  ...    -0.720442   
151          fwd           front   -0.513073 -1.263312  ...    -0.840335   
..           ...             ...         ...       ...  ...          ...   
139          fwd           front   -0.841690 -1.328231  ...    -0.456677   
143          fwd           front   -0.266610 -0.184033  ...    -0.456677   
77           fwd           front   -0.841690 -1.376921  ...    -0.840335   
146          fwd           front   -0.299471 -0.062309  ...    -0.456677   
39           fwd           front   -0.381626  0.091874  ...    -0.408720   

     fuel-system      bore    stroke compression-ratio  horsepower  peak-rpm  \
181         mpfi -0.230689  0.297653         -0.233162    1.399142  0.171291   
42          2bbl -0.677554  1.022393         -0.258667   -0.096240  0.798118   
36          1bbl -1.534045  0.486715         -0.233162   -0.737118  1.842832   
95          2bbl -0.677554  0.108590         -0.182151   -0.924041  0.171291   
151         2bbl -1.049941 -0.710682         -0.284172   -1.110964 -0.664480   
..           ...       ...       ...               ...         ...       ...   
139         2bbl  1.072667 -1.939589         -0.360688   -0.817228 -1.500251   
143         mpfi  1.072667 -1.939589         -0.284172   -0.256460  0.171291   
77          2bbl -1.347851 -0.080473         -0.182151   -0.950744  0.798118   
146         2bbl  1.072667 -1.939589         -0.284172   -0.576899 -0.664480   
39          1bbl -0.677554  1.022393         -0.284172   -0.470086  1.424946   

     city-mpg  highway-mpg    price  
181 -0.956587    -0.975215  15750.0  
42  -0.020386     0.053932  10345.0  
36   0.759781     0.494996   7295.0  
95   0.915815     0.936059   7799.0  
151  0.915815     1.083080   6338.0  
..        ...          ...      ...  
139  0.135647     0.053932   7053.0  
143  0.135647     0.200953   9960.0  
77   0.915815     1.083080   6189.0  
146  0.447714     0.200953   7463.0  
39   0.291681     0.347975   8845.0  

[199 rows x 25 columns]

In [14]:
# performing train-test split with 10% of the data as the test data
train_size = int(0.9 * len(data))
data_train = data_shuffle[:train_size].reset_index(drop=True)
data_test = data_shuffle[train_size:].reset_index(drop=True)

In [15]:
data_train.shape

(179, 25)

In [16]:
data_test.shape

(20, 25)

In [17]:
# separating the features and the target variable
X_train = data_train.drop('price', axis=1)
y_train = data_train.price
X_test = data_test.drop('price', axis=1)
y_test = data_test.price

## Q1. (a)
Train a k-nearest neighbors regression model, and report its validation set performance using root mean squared error.

## The model

In [18]:
def kNN(x):
    """Predicts the target variable for x using kNN regression for a given k."""
    # calculating the differences for the numerical variables except the target variable
    diff = X_train[cols_numerical[:-1]] - x[cols_numerical[:-1]]
    for col in cols_nominal:
        # calculating the dissimilarity for nominal variables and concatenating it to diff
        diff_col = (X_train[col] != x[col])
        diff = pd.concat([diff, diff_col], axis=1)
    diff = diff.astype(float)
    # calculating the Euclidean distance
    dist = lin.norm(diff, 2, axis=1)
    idx = np.argsort(dist)
    y_pred = 0
    # calculating the average of the target variable of the k nearest neighbours of x
    for i in range(k):
        y_pred += y_train[idx[i]]
    return y_pred / k

In [19]:
def kNN_predict():
    """Performs kNN regression on X_test and returns the predicted values."""
    return X_test.apply(kNN, axis=1)

In [20]:
def get_rmse(y_pred):
    """Calculates the root mean squared error of y_pred."""
    diff = y_test - y_pred
    return lin.norm(diff, 2)

### Training the model

In [21]:
# training a kNN regressor for k = 1 and calculating its RMSE
k = 1
y_pred = kNN_predict()
rmse = get_rmse(y_pred)
print(f'RMSE for kNN regressor with k = 1: {rmse}')

RMSE for kNN regressor with k = 1: 6004.161223684787


The root mean squared error for the k-nearest neighbours regression model on the validation set is 6004.16

## Q1. (b)
Find an optimal k for this model using cross-validation. 

### Optimizing the hyperparameters

In [22]:
# performing cross validation to find an optimal value for k
k_opt = 1
least_rmse = rmse
for k in range(1, 20, 2):
    y_pred = kNN_predict()
    rmse = get_rmse(y_pred)
    print(f'For k = {k}, RMSE = {rmse}')
    if rmse < least_rmse:
        least_rmse = rmse
        k_opt = k
print()
print(f'Optimal value of k: {k_opt}')

For k = 1, RMSE = 6004.161223684787
For k = 3, RMSE = 5475.784631752178
For k = 5, RMSE = 5076.139958669382
For k = 7, RMSE = 5533.313820554071
For k = 9, RMSE = 5476.131811533022
For k = 11, RMSE = 5601.011550463875
For k = 13, RMSE = 5902.802783273794
For k = 15, RMSE = 5637.139100248951
For k = 17, RMSE = 5144.668473519178
For k = 19, RMSE = 5446.281525689221

Optimal value of k: 5


The optimal k found for the model using cross validation is 5. 

## Q2. The census income dataset

In [23]:
# names of the columns of the dataset
cols = ['age', 'workclass', 'fnlwgt', 'education', 'education-num', 'marital-status', 'occupation', 'relationship', 'race', 'sex', 'capital-gain', 'capital-loss', 'hours-per-week', 'native-country', 'income']
cols

['age',
 'workclass',
 'fnlwgt',
 'education',
 'education-num',
 'marital-status',
 'occupation',
 'relationship',
 'race',
 'sex',
 'capital-gain',
 'capital-loss',
 'hours-per-week',
 'native-country',
 'income']

In [24]:
# reading the data
data = pd.read_csv('data_q2/adult.data', sep='\,\ ', engine='python', names=cols)
data

age         workclass  fnlwgt   education  education-num  \
0       39         State-gov   77516   Bachelors             13   
1       50  Self-emp-not-inc   83311   Bachelors             13   
2       38           Private  215646     HS-grad              9   
3       53           Private  234721        11th              7   
4       28           Private  338409   Bachelors             13   
...    ...               ...     ...         ...            ...   
32556   27           Private  257302  Assoc-acdm             12   
32557   40           Private  154374     HS-grad              9   
32558   58           Private  151910     HS-grad              9   
32559   22           Private  201490     HS-grad              9   
32560   52      Self-emp-inc  287927     HS-grad              9   

           marital-status         occupation   relationship   race     sex  \
0           Never-married       Adm-clerical  Not-in-family  White    Male   
1      Married-civ-spouse    Exec-managerial        Husband  White    Male   
2                Divorced  Handlers-cleaners  Not-in-family  White    Male   
3      Married-civ-spouse  Handlers-cleaners        Husband  Black    Male   
4      Married-civ-spouse     Prof-specialty           Wife  Black  Female   
...                   ...                ...            ...    ...     ...   
32556  Married-civ-spouse       Tech-support           Wife  White  Female   
32557  Married-civ-spouse  Machine-op-inspct        Husband  White    Male   
32558             Widowed       Adm-clerical      Unmarried  White  Female   
32559       Never-married       Adm-clerical      Own-child  White    Male   
32560  Married-civ-spouse    Exec-managerial           Wife  White  Female   

       capital-gain  capital-loss  hours-per-week native-country income  
0              2174             0              40  United-States  <=50K  
1                 0             0              13  United-States  <=50K  
2                 0             0              40  United-States  <=50K  
3                 0             0              40  United-States  <=50K  
4                 0             0              40           Cuba  <=50K  
...             ...           ...             ...            ...    ...  
32556             0             0              38  United-States  <=50K  
32557             0             0              40  United-States   >50K  
32558             0             0              40  United-States  <=50K  
32559             0             0              20  United-States  <=50K  
32560         15024             0              40  United-States   >50K  

[32561 rows x 15 columns]

## Data Preprocessing

### Analysing missing data

In [25]:
# putting NaN for missing values
data.replace('?', np.nan, inplace=True)
data

age         workclass  fnlwgt   education  education-num  \
0       39         State-gov   77516   Bachelors             13   
1       50  Self-emp-not-inc   83311   Bachelors             13   
2       38           Private  215646     HS-grad              9   
3       53           Private  234721        11th              7   
4       28           Private  338409   Bachelors             13   
...    ...               ...     ...         ...            ...   
32556   27           Private  257302  Assoc-acdm             12   
32557   40           Private  154374     HS-grad              9   
32558   58           Private  151910     HS-grad              9   
32559   22           Private  201490     HS-grad              9   
32560   52      Self-emp-inc  287927     HS-grad              9   

           marital-status         occupation   relationship   race     sex  \
0           Never-married       Adm-clerical  Not-in-family  White    Male   
1      Married-civ-spouse    Exec-managerial        Husband  White    Male   
2                Divorced  Handlers-cleaners  Not-in-family  White    Male   
3      Married-civ-spouse  Handlers-cleaners        Husband  Black    Male   
4      Married-civ-spouse     Prof-specialty           Wife  Black  Female   
...                   ...                ...            ...    ...     ...   
32556  Married-civ-spouse       Tech-support           Wife  White  Female   
32557  Married-civ-spouse  Machine-op-inspct        Husband  White    Male   
32558             Widowed       Adm-clerical      Unmarried  White  Female   
32559       Never-married       Adm-clerical      Own-child  White    Male   
32560  Married-civ-spouse    Exec-managerial           Wife  White  Female   

       capital-gain  capital-loss  hours-per-week native-country income  
0              2174             0              40  United-States  <=50K  
1                 0             0              13  United-States  <=50K  
2                 0             0              40  United-States  <=50K  
3                 0             0              40  United-States  <=50K  
4                 0             0              40           Cuba  <=50K  
...             ...           ...             ...            ...    ...  
32556             0             0              38  United-States  <=50K  
32557             0             0              40  United-States   >50K  
32558             0             0              40  United-States  <=50K  
32559             0             0              20  United-States  <=50K  
32560         15024             0              40  United-States   >50K  

[32561 rows x 15 columns]

In [26]:
# checking how many missing values are there in each column
data.isnull().sum()

age                  0
workclass         1836
fnlwgt               0
education            0
education-num        0
marital-status       0
occupation        1843
relationship         0
race                 0
sex                  0
capital-gain         0
capital-loss         0
hours-per-week       0
native-country     583
income               0
dtype: int64

In [27]:
# dropping the rows having missing data as we have a lot of data
data.dropna(inplace=True)
data

age         workclass  fnlwgt   education  education-num  \
0       39         State-gov   77516   Bachelors             13   
1       50  Self-emp-not-inc   83311   Bachelors             13   
2       38           Private  215646     HS-grad              9   
3       53           Private  234721        11th              7   
4       28           Private  338409   Bachelors             13   
...    ...               ...     ...         ...            ...   
32556   27           Private  257302  Assoc-acdm             12   
32557   40           Private  154374     HS-grad              9   
32558   58           Private  151910     HS-grad              9   
32559   22           Private  201490     HS-grad              9   
32560   52      Self-emp-inc  287927     HS-grad              9   

           marital-status         occupation   relationship   race     sex  \
0           Never-married       Adm-clerical  Not-in-family  White    Male   
1      Married-civ-spouse    Exec-managerial        Husband  White    Male   
2                Divorced  Handlers-cleaners  Not-in-family  White    Male   
3      Married-civ-spouse  Handlers-cleaners        Husband  Black    Male   
4      Married-civ-spouse     Prof-specialty           Wife  Black  Female   
...                   ...                ...            ...    ...     ...   
32556  Married-civ-spouse       Tech-support           Wife  White  Female   
32557  Married-civ-spouse  Machine-op-inspct        Husband  White    Male   
32558             Widowed       Adm-clerical      Unmarried  White  Female   
32559       Never-married       Adm-clerical      Own-child  White    Male   
32560  Married-civ-spouse    Exec-managerial           Wife  White  Female   

       capital-gain  capital-loss  hours-per-week native-country income  
0              2174             0              40  United-States  <=50K  
1                 0             0              13  United-States  <=50K  
2                 0             0              40  United-States  <=50K  
3                 0             0              40  United-States  <=50K  
4                 0             0              40           Cuba  <=50K  
...             ...           ...             ...            ...    ...  
32556             0             0              38  United-States  <=50K  
32557             0             0              40  United-States   >50K  
32558             0             0              40  United-States  <=50K  
32559             0             0              20  United-States  <=50K  
32560         15024             0              40  United-States   >50K  

[30162 rows x 15 columns]

In [28]:
# Replacing the target variable with 1 and -1 when the income is greater than and less than 50k respectively
data.income.replace('>50K', 1, inplace=True)
data.income.replace('<=50K', -1, inplace=True)
data

age         workclass  fnlwgt   education  education-num  \
0       39         State-gov   77516   Bachelors             13   
1       50  Self-emp-not-inc   83311   Bachelors             13   
2       38           Private  215646     HS-grad              9   
3       53           Private  234721        11th              7   
4       28           Private  338409   Bachelors             13   
...    ...               ...     ...         ...            ...   
32556   27           Private  257302  Assoc-acdm             12   
32557   40           Private  154374     HS-grad              9   
32558   58           Private  151910     HS-grad              9   
32559   22           Private  201490     HS-grad              9   
32560   52      Self-emp-inc  287927     HS-grad              9   

           marital-status         occupation   relationship   race     sex  \
0           Never-married       Adm-clerical  Not-in-family  White    Male   
1      Married-civ-spouse    Exec-managerial        Husband  White    Male   
2                Divorced  Handlers-cleaners  Not-in-family  White    Male   
3      Married-civ-spouse  Handlers-cleaners        Husband  Black    Male   
4      Married-civ-spouse     Prof-specialty           Wife  Black  Female   
...                   ...                ...            ...    ...     ...   
32556  Married-civ-spouse       Tech-support           Wife  White  Female   
32557  Married-civ-spouse  Machine-op-inspct        Husband  White    Male   
32558             Widowed       Adm-clerical      Unmarried  White  Female   
32559       Never-married       Adm-clerical      Own-child  White    Male   
32560  Married-civ-spouse    Exec-managerial           Wife  White  Female   

       capital-gain  capital-loss  hours-per-week native-country  income  
0              2174             0              40  United-States      -1  
1                 0             0              13  United-States      -1  
2                 0             0              40  United-States      -1  
3                 0             0              40  United-States      -1  
4                 0             0              40           Cuba      -1  
...             ...           ...             ...            ...     ...  
32556             0             0              38  United-States      -1  
32557             0             0              40  United-States       1  
32558             0             0              40  United-States      -1  
32559             0             0              20  United-States      -1  
32560         15024             0              40  United-States       1  

[30162 rows x 15 columns]

In [29]:
# checking the data types of the columns
data.dtypes

age                int64
workclass         object
fnlwgt             int64
education         object
education-num      int64
marital-status    object
occupation        object
relationship      object
race              object
sex               object
capital-gain       int64
capital-loss       int64
hours-per-week     int64
native-country    object
income             int64
dtype: object

### Label encoding the categorical variables

In [30]:
# label encoding the columns having nominal variables
cols_nominal = ['workclass', 'education', 'marital-status', 'occupation', 'relationship', 'race', 'sex', 'native-country']
for col in cols_nominal:
    data[col] = data[col].astype('category').cat.codes
data

age  workclass  fnlwgt  education  education-num  marital-status  \
0       39          5   77516          9             13               4   
1       50          4   83311          9             13               2   
2       38          2  215646         11              9               0   
3       53          2  234721          1              7               2   
4       28          2  338409          9             13               2   
...    ...        ...     ...        ...            ...             ...   
32556   27          2  257302          7             12               2   
32557   40          2  154374         11              9               2   
32558   58          2  151910         11              9               6   
32559   22          2  201490         11              9               4   
32560   52          3  287927         11              9               2   

       occupation  relationship  race  sex  capital-gain  capital-loss  \
0               0             1     4    1          2174             0   
1               3             0     4    1             0             0   
2               5             1     4    1             0             0   
3               5             0     2    1             0             0   
4               9             5     2    0             0             0   
...           ...           ...   ...  ...           ...           ...   
32556          12             5     4    0             0             0   
32557           6             0     4    1             0             0   
32558           0             4     4    0             0             0   
32559           0             3     4    1             0             0   
32560           3             5     4    0         15024             0   

       hours-per-week  native-country  income  
0                  40              38      -1  
1                  13              38      -1  
2                  40              38      -1  
3                  40              38      -1  
4                  40               4      -1  
...               ...             ...     ...  
32556              38              38      -1  
32557              40              38       1  
32558              40              38      -1  
32559              20              38      -1  
32560              40              38       1  

[30162 rows x 15 columns]

### Train-test split

In [31]:
# shuffling the rows of the dataframe
data_shuffle = data.sample(frac=1, random_state=1)
data_shuffle

age  workclass  fnlwgt  education  education-num  marital-status  \
12488   30          2  202450         11              9               2   
5130    40          2  217120          0              6               0   
8616    23          2  240398          9             13               4   
15453   47          5  100818         11              9               2   
13555   43          2  191814          7             12               2   
...    ...        ...     ...        ...            ...             ...   
18656   24          2  149457         15             10               4   
5661    19          2   93604          5              4               4   
13155   17          2  108402          0              6               4   
255     42          4  303044         11              9               2   
32100   26          2  245880         11              9               4   

       occupation  relationship  race  sex  capital-gain  capital-loss  \
12488          13             0     4    1             0             0   
5130            2             1     4    1             0             0   
8616           11             1     2    1             0             0   
15453           0             5     4    0             0             0   
13555          11             0     4    1          7688             0   
...           ...           ...   ...  ...           ...           ...   
18656           0             4     4    0             0             0   
5661            2             3     4    1             0          1602   
13155          11             3     4    0             0             0   
255             4             0     1    1             0             0   
32100           7             2     4    1             0             0   

       hours-per-week  native-country  income  
12488              65              38       1  
5130               50              38      -1  
8616               15              38      -1  
15453              40              38       1  
13555              50              38       1  
...               ...             ...     ...  
18656              40              38      -1  
5661               32              38      -1  
13155              20              38      -1  
255                40               0       1  
32100              40              38      -1  

[30162 rows x 15 columns]

In [32]:
# performing train-test split with 20% of the data as the test data
train_size = int(0.8 * len(data))
data_train = data_shuffle[:train_size].reset_index(drop=True)
data_test = data_shuffle[train_size:].reset_index(drop=True)

In [33]:
data_train.shape

(24129, 15)

In [34]:
data_test.shape

(6033, 15)

In [35]:
# separating the features and the target variable
X_train = data_train.drop('income', axis=1)
y_train = data_train.income
X_test = data_test.drop('income', axis=1)
y_test = data_test.income

## Q2. (a)
Train a decision tree classification model using information gain as the splitting criterion and using only single feature decision stumps at all non-leaf nodes and majority votes at leaf nodes, and report its validation set performance using % accuracy. 

## The model

In [36]:
class Node:
    # A node stores its own depth (root = depth 0), its decision stump, its parent and child information
    # Leaf nodes also store a constant label that is assigned to every data point that reaches that leaf
    def __init__(self, depth = 0, stump = (0,0), parent = None):
        self.depth = depth
        self.stump = stump
        self.parent = parent
        self.left = None
        self.right = None
        self.isLeaf = True
        self.label = 0
        
    def predict(self, data):
        # If it is a leaf node, it can predict rightaway
        if self.isLeaf:
            return self.label
        # Else, one of its children has to do the job
        else:
            if data[self.stump[0]] > self.stump[1]:
                return self.right.predict(data)
            else:
                return self.left.predict(data)
    
    # Gets the entropy of a node with nPos positive points and nNeg negative points
    def getEntropy(self, nPos, nNeg):
        if nPos == 0 or nNeg == 0:
            return 0
        nTot = nPos + nNeg
        # Finding the proportion of the positives and negatives in that node
        pPos = nPos / nTot
        pNeg = nNeg / nTot
        return - pPos * float(np.log(pPos)) - pNeg * float(np.log(pNeg))
            
    def getStump(self, X, y):
        n = y.size
        # Number of positive and negative labels at this node
        nPos = (y == 1).sum()
        nNeg = (y == -1).sum()
        bestObjective = -float('inf')
        
        # For each of the features in the data
        for i in range(X.shape[1]):
            # Do not use the same feature as used by the parent node
            if self.parent is not None and i == self.parent.stump[0]:
                continue
            # Finding out all values at which we can threshold that feature
            candidateThresholds = np.sort(X[:, i])
            idx = np.argsort(X[:, i])
            # The cumulative sum trick as the labels are binary 
            ySorted = y[idx]
            yCum = np.cumsum(ySorted)
            yCumRev = np.cumsum(ySorted[::-1])[::-1]
            # For each possible threshold (except the ones at the extreme)
            for j in range(1, candidateThresholds.size-1):
                # Calculating the information gain
                candidateObjective = self.getEntropy(nPos, nNeg) \
                                    - ((j + 1) / n) * self.getEntropy((yCum[j] + j + 1) / 2, (j + 1 - yCum[j]) / 2) \
                                    - ((n - j - 1) / n) * self.getEntropy((yCumRev[j+1] + n - j - 1) / 2, (n - j - 1 - yCumRev[j+1]) / 2)
                if candidateObjective > bestObjective:
                    bestObjective = candidateObjective
                    bestFeat = i
                    bestThresh = candidateThresholds[j]
        return bestFeat, bestThresh
            
    def train(self, X, y, maxLeafSize, maxDepth):
        # If too few data points are present, or else if this node is too deep in the tree, make this a leaf
        if y.size < maxLeafSize or self.depth >= maxDepth:
            self.isLeaf = True
            self.label = 1 if y.sum() >= 0 else -1
        else:
            # This node will be split and hence it is not a leaf
            self.isLeaf = False
            # Getting the best possible decision stump
            self.stump = self.getStump(X, y)
            self.left = Node(depth = self.depth + 1, parent = self)
            self.right = Node(depth = self.depth + 1, parent = self)
            # Finding which points go to the left child and which go to the right child
            discriminant = X[:, self.stump[0]] - self.stump[1]
            # Training the two children recursively
            self.left.train(X[discriminant <= 0, :], y[discriminant <= 0], maxLeafSize, maxDepth)
            self.right.train(X[discriminant > 0, :], y[discriminant > 0], maxLeafSize, maxDepth)

In [37]:
class Tree:
    def __init__(self, maxLeafSize = 10, maxDepth = 5):
        self.root = Node()
        self.maxLeafSize = maxLeafSize
        self.maxDepth = maxDepth 
        
    def predict(self, x):
        return self.root.predict(x)
    
    def train(self, X, y):
        self.root.train(X, y, self.maxLeafSize, self.maxDepth)

In [38]:
def dtree_predict(maxLeafSize, maxDepth):
    """Performs Decision Tree classification on X_test and returns the predicted labels."""
    tree = Tree(maxLeafSize, maxDepth)
    tree.train(np.array(X_train), np.array(y_train))
    return X_test.apply(tree.predict, axis=1)

In [39]:
def get_accuracy(y_pred):
    """Calculates the accuracy of y_pred."""
    y_pred = dtree_predict(maxLeafSize, maxDepth)
    return (y_pred == y_test).sum() / len(y_test) * 100

### Training the model

In [40]:
# training a Decision Tree classifier for maxLeafSize = 10 and maxDepth = 5 and calculating its accuracy
maxLeafSize = 10
maxDepth = 5
y_pred = dtree_predict(maxLeafSize, maxDepth)
accuracy = get_accuracy(y_pred)
print(f'Accuracy for Decision Tree classifier with max leaf size = {maxLeafSize} and max depth = {maxDepth}: {round(accuracy, 2)}%')

Accuracy for Decision Tree classifier with max leaf size = 10 and max depth = 5: 83.04%


The accuracy of the decision tree classification model on the validation set is 83.04%

## Q2. (b)
Use cross-validation to optimize the tree hyperparameters.

### Optimizing the hyperparameters

In [41]:
# performing cross validation to find optimal values for maxLeafSize and maxDepth
maxLeafSizeOpt = 10
maxDepthOpt = 5
max_accuracy = accuracy
for maxDepth in range(3, 7):
    for maxLeafSize in range(15, 0, -5):
        y_pred = dtree_predict(maxLeafSize, maxDepth)
        accuracy = get_accuracy(y_pred)
        print(f'For max leaf size = {maxLeafSize} and max depth = {maxDepth}, accuracy = {round(accuracy, 2)}%')
        if accuracy > max_accuracy:
            max_accuracy = accuracy
            maxLeafSizeOpt = maxLeafSize
            maxDepthOpt = maxDepth
print()
print(f'Optimal value of maxLeafSize: {maxLeafSizeOpt}')
print(f'Optimal value of maxDepth: {maxDepthOpt}')

For max leaf size = 15 and max depth = 3, accuracy = 78.24%
For max leaf size = 10 and max depth = 3, accuracy = 78.24%
For max leaf size = 5 and max depth = 3, accuracy = 78.24%
For max leaf size = 15 and max depth = 4, accuracy = 78.7%
For max leaf size = 10 and max depth = 4, accuracy = 78.7%
For max leaf size = 5 and max depth = 4, accuracy = 78.7%
For max leaf size = 15 and max depth = 5, accuracy = 83.04%
For max leaf size = 10 and max depth = 5, accuracy = 83.04%
For max leaf size = 5 and max depth = 5, accuracy = 83.04%
For max leaf size = 15 and max depth = 6, accuracy = 82.89%
For max leaf size = 10 and max depth = 6, accuracy = 82.89%
For max leaf size = 5 and max depth = 6, accuracy = 82.91%

Optimal value of maxLeafSize: 10
Optimal value of maxDepth: 5


The optimal values of the tree hyperparameters found using cross validation are:
- Maximum leaf size: 10
- Maximum depth of the tree: 5